src/nvim/api/vim.c

```diff
@@ -207,18 +208,38 @@ String nvim_replace_termcodes(String str, Boolean from_part, Boolean do_lt,
   return cstr_as_string(ptr);
 }
 
-String nvim_command_output(String str, Error *err)
+/// Executes an ex-command and returns its (non-error) output.
+/// Shell |:!| output is not captured.
+///
+/// On parse error: forwards the Vim error; does not update v:errmsg.
+/// On runtime error: forwards the Vim error; does not update v:errmsg.
+///
+/// @param command  Ex-command string
+/// @param[out] err Error details (Vim error), if any
+String nvim_command_output(String command, Error *err)
   FUNC_API_SINCE(1)
 {
-  do_cmdline_cmd("redir => v:command_output");
-  nvim_command(str, err);
-  do_cmdline_cmd("redir END");
+  Array args = ARRAY_DICT_INIT;
+  ADD(args, STRING_OBJ(command));
+  ADD(args, STRING_OBJ(cstr_as_string("silent")));
 
+  Object rv = nvim_call_function(cstr_as_string("execute"), args, err);
```

Comment:
are there much reason to not use `f_execute` directly, or (maybe even better) factor out common implementation function?

Reply:
One reason is I've already spent too much time on this :)

Reply:
Another reason is `nvim_call_function` is already taking care of the try_start dance, etc.

Reply:
Creating typvals directly is simpler and cleared than first creating API values only to be converted to typvals. So you replaced a meaningful (and standard pattern) dance with a meaningless dance.

Reply:
It's not clearer to me. And I thought we had planned to use the API internally?

Reply:
Calling a vimscript function indirectly to implement an API function is definitely not "using the API internally" (that design pattern would be rather defining `execute()` in terms of an API function)
If one for some reason needs to construct an explicit typval there is no reason to construct an API _value_ and immediately convecting it to typval. This just moves _horizontally_ between different abstractions, at a performance cost, and not across whatever internal/external boundary. Not that performance matters matters in this particular case, but it is an anti-pattern to be avoided. ( i e "performance as a consequence of doing things right" and not an optimization)

Reply:
I agree, it would be cleaner to make `f_execute` call API `nvim_command_output()` and not the other way around. Though given how “much” code is actually needed I would rather go with calling `do_cmdline` directly with a custom `LineGetter`: basically you need

```

{
  garray_T *const save_capture_ga = &capture_ga;
  garray_T capture_local;
  ga_init(&capture_local, 1, 80);
  capture_ga = &capture_local;
  try_start();
  do_cmdline(NULL, get_string_line, &string,
              DOCMD_NOWAIT|DOCMD_VERBOSE|DOCMD_REPEAT|DOCMD_KEYTYPED);
  try_end(&err);
  Object ret = STRING_OBJ(cstr_as_string(capture_local.ga_data));
  capture_ga = save_capture_ga;
  return ret;
}
```

BTW, `f_execute` actually needs neither `vim_strsave()` nor `ga_clear()`, that is only useless waste of time.

Reply:
Plus actual get_string_line, but that is not hard either.

Reply:
@ZyX-I if we go this route, it might make sense to define `nvim_source` which sources an API list using custom reader and also has output/error handling controlled by parameter.

Reply:
@ZyX-I Why is `get_string_line` variant needed? `nvim_command_output` only accepts String and returns String, so I guess `do_cmdline_cmd(command.data)` should be enough.